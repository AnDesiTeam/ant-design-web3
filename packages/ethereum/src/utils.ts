import debug from 'debug';

const createDebug = (namespace: string) => {
  return debug(`antd-web3:ethereum:${namespace}`);
};

export { createDebug };
