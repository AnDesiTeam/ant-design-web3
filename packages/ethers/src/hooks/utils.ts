import * as ethers from 'ethers';
import * as ethers5 from 'ethers5';

export const isEthersV5 = () => ethers.version && ethers.version.startsWith('5.');

export const FallbackProvider = isEthersV5()
  ? ethers5.providers.FallbackProvider
  : ethers.FallbackProvider;

export const JsonRpcProvider = isEthersV5()
  ? ethers5.providers.JsonRpcProvider
  : ethers.JsonRpcProvider;

export const BrowserProvider = isEthersV5()
  ? ethers5.providers.Web3Provider
  : ethers.BrowserProvider;
