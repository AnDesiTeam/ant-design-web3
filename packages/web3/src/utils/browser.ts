export const getPlatform = () => {
  const userAgent = navigator.userAgent.toLowerCase();
  if (userAgent.includes('chrome')) {
    return 'Chrome';
  } else if (userAgent.includes('firefox')) {
    return 'Firefox';
  } else if (userAgent.includes('edge')) {
    return 'Edge';
  } else if (userAgent.includes('safari')) {
    return 'Safari';
  } else {
    return 'Other';
  }
};

export const writeCopyText = (text: string): Promise<void> => {
  return navigator.clipboard.writeText(text);
};

export const readCopyText = (): Promise<string> => {
  return navigator.clipboard.readText();
};
