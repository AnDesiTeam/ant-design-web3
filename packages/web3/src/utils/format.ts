export const fillWith0x = (address: string = ''): string => {
  const filledAddress = address.startsWith('0x') ? address : `0x${address}`;
  return filledAddress;
};

export const formatAddress = (address: string = '', groupSize = 4): string => {
  const formattedGroups = [];

  for (let i = 0; i < address.length; i += groupSize) {
    const group = address.slice(i, i + groupSize);
    formattedGroups.push(group);
  }

  return formattedGroups.join(' ');
};
