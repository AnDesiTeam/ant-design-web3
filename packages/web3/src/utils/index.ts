export const getPlatform = () => {
  const userAgent = navigator.userAgent.toLowerCase();

  if (userAgent.includes('chrome')) {
    return 'Chrome';
  } else if (userAgent.includes('firefox')) {
    return 'Firefox';
  } else if (userAgent.includes('edge')) {
    return 'Edge';
  } else if (userAgent.includes('safari')) {
    return 'Safari';
  } else {
    return 'Other';
  }
};
