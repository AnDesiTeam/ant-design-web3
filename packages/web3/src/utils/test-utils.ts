import { vi } from 'vitest';

const oriClipboard = window.navigator.clipboard;
// 代理 navigator.clipboard 方法用于测试复制文本到粘贴板的功能
export const mockClipboard = () => {
  const clipboard = {
    text: '',
    writeText: (text: string) => {
      clipboard.text = text;
      return Promise.resolve();
    },
    readText: () => {
      return Promise.resolve(clipboard.text);
    },
  };
  Object.defineProperty(window, 'navigator', {
    value: {
      clipboard,
    },
  });

  return () => {
    Object.defineProperty(window, 'navigator', {
      value: {
        clipboard: oriClipboard,
      },
    });
  };
};

/**
 * Wait for a time delay. Will wait `advanceTime * times` ms.
 *
 * @param advanceTime Default 1000
 * @param times Default 20
 */
export async function waitFakeTimer(advanceTime = 1000, times = 20) {
  for (let i = 0; i < times; i += 1) {
    // eslint-disable-next-line no-await-in-loop
    await vi.waitFor(async () => {
      await Promise.resolve();

      if (advanceTime > 0) {
        vi.advanceTimersByTime(advanceTime);
      } else {
        vi.runAllTimers();
      }
    });
  }
}
