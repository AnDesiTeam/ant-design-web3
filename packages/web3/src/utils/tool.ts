import type { GlobalToken } from 'antd';
import { TinyColor } from '@ctrl/tinycolor';

export const isDarkTheme = (token: GlobalToken) => {
  const hsv = new TinyColor(token.colorFill).onBackground(token.colorBgElevated).toHsv();
  return hsv.v < 0.5;
};

export const formatNumUnit = (num: number) => {
  return num > 1000 ? `${(num / 1000).toFixed(1)}k` : num;
};
