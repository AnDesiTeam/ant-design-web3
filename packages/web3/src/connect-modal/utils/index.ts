import type React from 'react';
import type { Wallet } from '../interface';
import { getPlatform } from '../../utils';

export const mergeReactNodeProps = (
  node: React.ReactNode,
  defaultNode: React.ReactNode,
  resolveString?: (node: string) => React.ReactNode,
) => {
  if (node === null || node === false) return node;
  if (typeof node === 'string' && resolveString) return resolveString(node);
  return node ?? defaultNode;
};

export const defaultGroupOrder = (a: string, b: string) => a.localeCompare(b);

export const getWalletRoute = (wallet: Wallet) => {
  const curExtensions =
    wallet?.extensions && wallet.extensions.some((item) => item.key === getPlatform());

  if (curExtensions) {
    return 'wallet';
  }

  if (wallet?.app) {
    return 'qrCode';
  }

  return 'unknown';
};
