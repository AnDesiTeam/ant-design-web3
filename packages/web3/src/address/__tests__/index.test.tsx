import { Address } from '..';
import { fireEvent, render } from '@testing-library/react';
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { mockClipboard } from '../../utils/test-utils';
import { readCopyText } from '../../utils';

describe('Address', () => {
  let resetMockClipboard: () => void;
  beforeEach(() => {
    resetMockClipboard = mockClipboard();
  });
  afterEach(() => {
    resetMockClipboard();
  });

  it('mount correctly', () => {
    expect(() => render(<Address />)).not.toThrow();
  });

  it('display address', () => {
    const { baseElement } = render(
      <Address address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B" />,
    );
    expect(baseElement.querySelector('.ant-web3-address')?.textContent).toBe(
      '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
    );
  });

  it('display address with ellipsis', () => {
    const { baseElement } = render(
      <Address ellipsis address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B" />,
    );
    expect(baseElement.querySelector('.ant-web3-address')?.textContent).toBe('0x21CD...Fd3B');
  });

  it('display address with ellipsis and custom clip', () => {
    const { baseElement } = render(
      <Address
        ellipsis={{
          headClip: 3,
          tailClip: 3,
        }}
        address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B"
      />,
    );
    expect(baseElement.querySelector('.ant-web3-address')?.textContent).toBe('0x2...d3B');
  });

  it('display address with tooltip', async () => {
    const { baseElement } = render(
      <Address address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B" ellipsis />,
    );

    expect(baseElement.querySelector('.ant-web3-address')?.textContent).toBe('0x21CD...Fd3B');
    fireEvent.mouseEnter(baseElement.querySelector('.ant-web3-address-text')!);
    await vi.waitFor(() => {
      expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent).toBe(
        '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
      );
    });
  });

  it('display address with default format', () => {
    const { baseElement } = render(
      <Address address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B" format />,
    );
    expect(baseElement.querySelector('.ant-web3-address')?.textContent).toBe(
      '0x 21CD f097 4d53 a6e9 6eF0 5d7B 324a 9803 735f Fd3B',
    );
    expect(baseElement.querySelector('.ant-web3-address')).toMatchSnapshot();
  });

  it('display address with custom format', () => {
    const { baseElement } = render(
      <Address
        address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B"
        format={(input) => {
          return input.slice(0, 10);
        }}
      />,
    );
    expect(baseElement.querySelector('.ant-web3-address')?.textContent).toBe('0x21CDf097');
  });
  it('display address with copyable', async () => {
    const { baseElement } = render(
      <Address address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B" ellipsis copyable />,
    );
    expect(baseElement.querySelector('.ant-web3-address')?.textContent).toBe('0x21CD...Fd3B');
    fireEvent.click(baseElement.querySelector('.anticon-copy')!);
    await vi.waitFor(() => {
      expect(baseElement.querySelector('.ant-message')).not.toBeNull();
      expect(baseElement.querySelector('.ant-message-notice-content')?.textContent?.trim()).toBe(
        'Address Copied!',
      );
      expect(readCopyText()).resolves.toBe('0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B');
    });
  });
  it('does not display ellipsis if ellipsis is set to false', () => {
    const { baseElement } = render(
      <Address ellipsis={false} address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B" />,
    );
    expect(baseElement.querySelector('.ant-web3-address')?.textContent).toBe(
      '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
    );
  });
  it('does not display tooltip if tooltip is set to false', async () => {
    const { baseElement } = render(
      <Address tooltip={false} address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B" />,
    );
    fireEvent.mouseEnter(baseElement.querySelector('.ant-web3-address-text')!);
    await vi.waitFor(() => {
      expect(baseElement.querySelector('.ant-tooltip-inner')).toBeNull();
    });
  });
  it('does not apply any formatting if format is set to false', () => {
    const { baseElement } = render(
      <Address format={false} address="0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B" />,
    );
    expect(baseElement.querySelector('.ant-web3-address')?.textContent).toBe(
      '0x21CDf0974d53a6e96eF05d7B324a9803735fFd3B',
    );
  });
});
