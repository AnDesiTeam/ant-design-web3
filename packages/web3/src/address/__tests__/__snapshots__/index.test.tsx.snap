// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Address > display address with default format 1`] = `
<div
  class="ant-space css-dev-only-do-not-override-1ck3jst ant-space-horizontal ant-space-align-center ant-space-gap-row-small ant-space-gap-col-small ant-web3-address css-dev-only-do-not-override-1ck3jst"
>
  <div
    class="ant-space-item"
  >
    <span
      class="ant-web3-address-text"
    >
      0x21 CDf0 974d 53a6 e96e F05d 7B32 4a98 0373 5fFd 3B
    </span>
  </div>
</div>
`;
